casper/src/main/scala/io/casperlabs/casper/FinalityDetectorBySingleSweep.scala

```
@@ -0,0 +1,341 @@
+          } yield (updatedBlockLevelTags, updatedValidatorLevel)
+      }
+  }
+
+  /*
+   * Finds the best level-1 committee for a given candidate block
+   */
+  def findBestCommittee(
+      dag: DagRepresentation[F],
+      candidateBlockHash: BlockHash,
+      weights: Map[Validator, Long]
+  ): F[Option[Committee]] =
+    for {
+      committeeApproximationOpt <- FinalityDetectorUtil.committeeApproximation[F](
+                                    dag,
+                                    candidateBlockHash,
+                                    weights
+                                  )
+      result <- committeeApproximationOpt match {
+                 case Some((committeeApproximation, maxWeightApproximation)) =>
+                   for {
+                     levelZeroMsgs <- levelZeroMsgs(
+                                       dag,
+                                       candidateBlockHash,
+                                       committeeApproximation
+                                     )
+                     committeeMembersAfterPruning <- findBestQLoop(
+                                                      dag,
+                                                      committeeApproximation.toSet,
+                                                      levelZeroMsgs,
+                                                      weights,
+                                                      maxWeightApproximation
+                                                    )
+                   } yield committeeMembersAfterPruning
+                 case None =>
+                   none[Committee].pure[F]
+               }
+    } yield result
+
+  override def onNewBlockAddedToTheBlockDag(
+      dag: DagRepresentation[F],
+      block: Block,
+      latestFinalizedBlock: BlockHash
+  ): F[Unit] = ().pure[F]
+
+  override def rebuildFromLatestFinalizedBlock(
+      dag: DagRepresentation[F],
+      newFinalizedBlock: BlockHash
+  ): F[Unit] = ().pure[F]
```

This is confusing. I understand that different versions of finality detector answer different questions. We should at least raise an error here rather than pretend everything worked.

---

Done